server/src/main/java/org/opensearch/index/translog/transfer/BlobStoreTransferService.java

```diff
@@ -62,14 +74,98 @@ public void uploadBlobAsync(
     }
 
     @Override
-    public void uploadBlob(final TransferFileSnapshot fileSnapshot, Iterable<String> remoteTransferPath) throws IOException {
-        assert remoteTransferPath instanceof BlobPath;
+    public void uploadBlob(final TransferFileSnapshot fileSnapshot, Iterable<String> remoteTransferPath, WritePriority writePriority)
+        throws IOException {
         BlobPath blobPath = (BlobPath) remoteTransferPath;
         try (InputStream inputStream = fileSnapshot.inputStream()) {
             blobStore.blobContainer(blobPath).writeBlobAtomic(fileSnapshot.getName(), inputStream, fileSnapshot.getContentLength(), true);
         }
     }
 
+    @Override
+    public void uploadBlobs(
+        Set<TransferFileSnapshot> fileSnapshots,
+        final Map<Long, BlobPath> blobPaths,
+        ActionListener<TransferFileSnapshot> listener,
+        WritePriority writePriority
+    ) {
+        List<CompletableFuture<Void>> resultFutures = new ArrayList<>();
+        fileSnapshots.forEach(fileSnapshot -> {
+            BlobPath blobPath = blobPaths.get(fileSnapshot.getPrimaryTerm());
+            if (!(blobStore.blobContainer(blobPath) instanceof VerifyingMultiStreamBlobContainer)) {
+                uploadBlobByThreadPool(ThreadPool.Names.TRANSLOG_TRANSFER, fileSnapshot, blobPath, listener, writePriority);
+            } else {
+                CompletableFuture<Void> resultFuture = createUploadFuture(fileSnapshot, listener, blobPath, writePriority);
+                if (resultFuture != null) {
+                    resultFutures.add(resultFuture);
+                }
+            }
+        });
+
+        if (resultFutures.isEmpty() == false) {
+            CompletableFuture<Void> resultFuture = CompletableFuture.allOf(resultFutures.toArray(new CompletableFuture[0]));
+            try {
+                resultFuture.get();
+            } catch (Exception e) {
+                logger.warn("Failed to upload blobs", e);
+            }
+        }
```

Comment:
We already have a LatchedActionListener that is awaiting upload https://github.com/opensearch-project/OpenSearch/blob/240193a8b129aaffffb39a57403264b8c4893efb/server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java#L130-L140 

Do we really need to block on this?

Reply:
CompletableFuture needs a get or a join to start executing. This can be further pushed up though but there will always be one thread waiting for all the chained or merged futures.

Reply:
>CompletableFuture needs a get or a join to start executing

I don't think this is true. A get or a join is needed to block on result and not to start the execution.